src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/ControllerExtensions.cs

```diff
@@ -0,0 +1,83 @@
+using System;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Localization;
+using OrchardCore.Data;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Tenants.ViewModels;
+
+namespace OrchardCore.Tenants.Controllers
+{
+    internal static class ControllerExtensions
+    {
+        internal static async Task ValidateModelAsync<TViewModel>(this Controller controller, TViewModel model, bool newTenant = true) where TViewModel : TenantViewModel
+        {
+            var shellHost = controller.HttpContext.RequestServices.GetService<IShellHost>();
+            var featureProfilesService = controller.HttpContext.RequestServices.GetService<IFeatureProfilesService>();
+            var databaseProviders = controller.HttpContext.RequestServices.GetServices<DatabaseProvider>();
+            var S = controller.HttpContext.RequestServices.GetService<IStringLocalizer<TViewModel>>();
```

Comment:
what will the po extractor discover these as?

Does it break all the existing translations or not?

Reply:
> what will the po extractor discover these as?

Probably yes coz he looks for `S`, `H` and `T` /cc @lukaskabrt 

> Does it break all the existing translations or not?

Why it should break the existing translations?!!

Reply:
> what will the po extractor discover these as?

Probably yes coz he looks for `S`, `H` and `T` /cc @lukaskabrt 

> Does it break all the existing translations or not?

Why it should break the existing translations?!!

Reply:
https://github.com/lukaskabrt/PoExtractor/blob/2ebda87de7ece8f61edd0cb1042915f9823ed00f/PoExtractor.DotNet.CS/SingularStringExtractor.cs#L23

This line should handle any identifier has name of one of the `LocalizerIdentifiers`

Reply:
It breaks all existing translations because they are now in a different class, so wil lbe contexualised under that class.

It doesn't work because you resolve `var S = controller.HttpContext.RequestServices.GetService<IStringLocalizer<TViewModel>>();`

but the translations are contexualised in the new class, `ControllerExtensions`.
So to resolve them correctly you need to resolve a localizer for the extensions class.

Reply:
I got what you mean now, because the resources are discovered in the `ControllerExtensions` class

Reply:
Static types cannot be used as type arguments

Reply:
You won't be able to use a localizer in an extension method then (without dificult hacks)
You'll have to make it a service (better anyway to be honest).

Reply:
> You won't be able to use a localizer in an extension method then (without dificult hacks)

What about resolving `IStringLocalizer`? Something similar to `ILogger` in users module?

Reply:
@Piedone can we make this simple as it was, then think about simple & elegent solution to be applied in all the Controllers & ApiController accross the module. Seems this take much time than fixing the actual bug ;)

Also the localizer missing here

https://github.com/OrchardCMS/OrchardCore/blob/f6d77147885529ff9889b9709ffc078bdeddaf72/src/OrchardCore.Modules/OrchardCore.Users/Controllers/ControllerExtensions.cs#L86

perhaps for the same reason

Reply:
Why not just move the logic out to a service? I understand where you're coming from to just concentrate on the bug but we should also strive to continuously refactactor as necessary, and in this case, we'd add to the pile of copy-pasted code (that already had a bug because of that and as your example shows, if anybody touches it they'll introduce new ones). So, the best time to do this was when this was originally written, the second best time is now :).